src/exposurelog/routers/find_exposures.py

```diff
@@ -0,0 +1,241 @@
+            begin=astropy_from_datetime(min_date),
+            end=astropy_from_datetime(max_date),
+        )
+        conditions.append("exposure.timespan OVERLAPS date_span")
+
+    # If order_by does not include "id" then append it, to make the order
+    # repeatable. Otherwise different calls can return data in different
+    # orders, which is a disaster when using limit and offset.
+    where = " and ".join(conditions)
+
+    loop = asyncio.get_running_loop()
+    rows = await loop.run_in_executor(
+        None,
+        find_exposures_in_registries,
+        state.registries,
+        instrument,
+        bind,
+        where,
+        limit,
+    )
+
+    exposures = [Exposure(**dict_from_exposure(row)) for row in rows]
+    return sorted(exposures, key=lambda exposure: exposure.obs_id)
+
+
+def astropy_from_datetime(
+    date: typing.Optional[datetime.datetime],
+) -> typing.Optional[astropy.time.Time]:
+    """Convert an optional TAI datetime.datetime to an astropy.time.Time.
+
+    Return None if date is None.
+    """
+    if date is None:
+        return None
+    return astropy.time.Time(date, scale="tai")
+
+
+def dict_from_exposure(
+    exposure: lsst.daf.butler.core.DimensionRecord,
+) -> dict:
+    data = exposure.toDict()
+    timespan = data.pop("timespan")
+    data["timespan_begin"] = timespan.begin.datetime
+    data["timespan_end"] = timespan.end.datetime
+    return data
+
+
+def find_exposures_in_registries(
+    registries: typing.Sequence[lsst.daf.butler.Registry],
+    instrument: str,
+    bind: dict,
+    where: str,
+    limit: int = 50,
+) -> typing.ValuesView[lsst.daf.butler.core.DimensionRecord]:
+    """Find exposures matching specified criteria.
+
+    The exposures are sorted by obs_id.
+
+    Parameters
+    ----------
+    registries
+        One or more data registries.
+        They are searched in order.
+    instrument
+        Name of instrument.
+    bind
+        bind argument to `lsst.daf.butler.Registry.queryDimensionRecords`.
+    where
+        where argument to `lsst.daf.butler.Registry.queryDimensionRecords`.
+    limit
+        Maximum number of exposures to return.
+
+    Returns
+    -------
+    exposures
+        The matching exposures.
+
+    Notes
+    -----
+    There is no pagination support because daf_butler does not support it.
+    The limit argument merely prevents performance issues from overly
+    broad searches.
+    """
+    # Keep records in a dict to avoid duplicates
+    # (which will only be a problem if we have two registries).
+    record_dict: typing.Dict[
+        str, lsst.daf.butler.core.DimensionRecord
+    ] = dict()
+    for registry in registries:
+        if len(record_dict) >= limit:
+            break
+        try:
+            record_iter = registry.queryDimensionRecords(
+                "exposure",
+                instrument=instrument,
+                bind=bind,
+                where=where,
+            )
```

Comment:
It is possible for this method to be implemented as a generator, and hence wouldn't do anything all until you start iteration over the results (e.g. raise that exception you're looking to catch).  I don't think that's the case right now, but it may not be safe to assume that remains true.